### PR TITLE
A more generic Polling middleware

### PR DESCRIPTION
**Pull request by boaz0**
//cc @karelhala - mind having a look? :smile:

### CODE REVIEW

**Review by karelhala — APPROVED** [review]
Looking hella good! Just a few minor enhancements and we are golden!

File: src/store/polling.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+const POLLING_SUFFIX = 'POLLING';
+
+const hasPollingData = (polling = {}) => {
+  return (
+    polling?.id &&
+    polling?.fetcher &&
+    polling?.condition &&
+    polling?.notifications
```
[the remark this answers is not in the document]
Comment (boaz0) [reply src/store/polling.js:8]: That's correct. I will remove this.

File: src/store/polling.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  );
+};
+
+const ofPollingType = (type) => {
+  return type.endsWith(`_${POLLING_SUFFIX}`);
+};
```
[the remark this answers is not in the document]
Comment (boaz0) [reply src/store/polling.js]: Done

File: src/store/polling.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  const { type, payload, meta } = action;
+  if (ofPollingType(type) && hasPollingData(payload)) {
+    const { id, fetcher, condition, notifications } = payload;
+    if (pollHash[id]) {
+      return;
+    }
+    pollHash[id] = true;
+    setTimeout(() => {
+      dispatch({
+        type: `FETCH_${id}`,
```
[the remark this answers is not in the document]
Comment (boaz0) [reply src/store/polling.js]: DONE :+1:

File: src/store/polling.js [hunk 4]
```diff
@@ -1,25 +1,50 @@
...
+    }
+    pollHash[id] = true;
+    setTimeout(() => {
+      dispatch({
+        type: id,
+        payload: fetcher().then((resp) => {
+          pollHash[id] = false;
+          const [toContinue, stateName] = condition(resp);
+          const nextActions = onEvent?.[stateName] || [];
+          nextActions.forEach((action) => action(dispatch));
```
Comment (boaz0) [inline src/store/polling.js:26]: @karelhala - please notice that I moved from `notifications` to `onEvent` that its an object where the key is the event name (`success`, `failure`, etc) and the value is an array of functions that take `dispatch` and use it to dispatch an action.

Is that OK?

Comment (karelhala) [reply src/store/polling.js:26]: Yeah, this is much better!

**Review by karelhala — APPROVED** [review]

**Comment by boaz0** [comment]
Thanks a lot @karelhala !!!

**Pull request merged**